src/SquidMath.h

```diff
@@ -123,5 +150,66 @@ SetToNaturalSumOrMax(S &var, Args... args)
     return var;
 }
 
+// If NaturalProduct() performance becomes important, consider using GCC and clang
+// built-ins like __builtin_mul_overflow() instead of manual overflow checks.
+
+/// \returns an exact, non-overflowing product of the arguments (or nothing)
+/// \returns nothing if at least one of the arguments is negative
+template <typename T, typename U>
+Optional<T>
+IncreaseProduct(T t, U u)
+{
+    static_assert(ValidateTypeTraits<T>(), "the first argument has a valid type");
+    static_assert(ValidateTypeTraits<U>(), "the second argument has a valid type");
+
+    // assume that callers treat negative numbers specially (see IncreaseSum() for details)
+    if (Less(t, 0) || Less(u, 0))
+        return Optional<T>();
+
+    // check that both operands do not overflow the result type
+    //if (!IncreaseSum(P(0), t).has_value() || !IncreaseSum(P(0), u).has_value())
+    //    return Optional<T>();
```

Comment:
In 1fa427f I tried to apply to Product methods the same design used already in Sum methods. The Sum methods do not check that each of its arguments fits into the result type - I think they work correctly without this check. The Product version should also work correctly - all possible overflows should be checked below. The only scenario why we could still need these checks are outlined in the three (commented) unit tests:
```
//TestSuccessForFirstMultiplicationType(max64u, zero8u);
//TestSuccessForFirstMultiplicationType(max64u, zero8s);
//TestSuccessForFirstMultiplicationType(max64u, zero64s);
```
The tests require that expressions such as `uint8_t=zero8u*max64u` overflow but the current IncreaseProduct() returns zero without overflowing. We should decide what to do in this case: either treat zero operands as 'overflow-free' operation (as it is now) or uncomment the check and the unit tests.

Reply:
I cannot think of a good reason to declare an overflow when multiplying by zero if/when doing so requires _more_ Product code.

AFAICT, Sum (and now Product) should follow several rules, including:

1. No operation _performed_  by the code may result in Undefined Behavior. According to C++ standard, _signed_ integer overflows or underflows may result in UB. This principle applies to every operation including every addition, multiplication, subtraction, division, and assignment.

2. Any negative operands must result in a valueless Optional<T> final result[^1].

3. Otherwise, the final result must be an Optional<T> object containing the exact/correct sum/product value (from basic math point of view) if that value can be represented by T.

4. Otherwise, the final result must be a valueless Optional<T>.

[^1]: Here, T is the requested underlying final result type.

If we can correctly satisfy these rules when dealing with zero operands, then let's completely remove the commented out code this change request is attached to and uncomment/adjust the corresponding test cases accordingly. Otherwise, let's discuss which rules are difficult to support when dealing with zero operands.